src/disk_util.hpp

```diff
@@ -0,0 +1,189 @@
+// Copyright 2018 Chia Network Inc
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//    http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef SRC_CPP_DISK_UTIL_HPP_
+#define SRC_CPP_DISK_UTIL_HPP_
+
+#include <string>
+#include <sstream>
+#include <fstream>
+#include <iostream>
+
+#ifndef _WIN32
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/file.h>
+#include <unistd.h>
+#include <limits.h>
+#include <stdlib.h>
+#endif
+
+#if !defined(_WIN32) && !defined(__APPLE__)
+#include <sys/sysmacros.h>
+#endif
+
+#include "util.hpp"
+
+namespace DiskUtil {
+
+    // For SSDs parallel writing is preferred.
+    // For HDDs it's slow and causes fragmentation.
+    static inline bool IsParallelWritingPreferred(const std::string &dir)
+    {
+#if defined(__APPLE__) || defined(_WIN32)
+        return true;
+#else
+        struct stat s;
+        memset(&s, 0, sizeof(s));
+
+        if (0 != stat(dir.c_str(), &s)) {
+            std::ostringstream err;
+            err << "Unable to find device name for dir " << dir << ": "
+                << strerror(errno);
+            throw std::runtime_error(err.str());
```

Comment:
it seems a bit gratuitous to throw here. The only thing you do with this exception is to turn it into a `false`. you might as well just return `false` directly (unless I'm misreading something)

Reply:
or `true` I suppose here, whatever is the current behavior

Reply:
I used exception because it's error path. Directory absence is unexpected. Also, errors from these exceptions are being logged and can be useful, i.e. "File not found" or "Access denied" etc.

Reply:
@arvidn If you'd like to avoid exceptions in this function please let me know, I fixed / pushed the rest.

Reply:
I realized that error handling is inconsistent with Lock/UnlockDir, I'll refactor this func to avoid exceptions.

Reply:
Done. Pushed the fixes.